File: Codes in Python3/Dynamic Programming/0-1 Knapsack.py
from os import *
from sys import *
from collections import *
from math import *
def knapsack(weights, values, n, W):
    # Initialize the DP array for the previous row with 0 values
    prev = [0] * (W + 1)

    # Base case initialization: for the first item
    for i in range(weights[0], W + 1):
        prev[i] = values[0]

    # Iterate through all items
    for ind in range(1, n):
        # To avoid overwriting values we still need to read, iterate backwards
        for weight in range(W, weights[ind] - 1, -1):
            # If including the item is possible, and yields a better value, include it
            prev[weight] = max(prev[weight], values[ind] + prev[weight - weights[ind]])

    # The last element in 'prev' now represents the maximum value achievable
    return prev[W]

# Read input as specified in the question
T = int(input())  # Read number of test cases

for _ in range(T):
    N = int(input())  # Number of items
    weights = list(map(int, input().split()))  # Weights of the items
    values = list(map(int, input().split()))  # Values of the items
    W = int(input())  # Maximum weight of the knapsack

    # Call the knapsack function and print the result
    print(knapsack(weights, values, N, W))


# def maxProfit(values, weights, n, w):
#     KS=[[0] * (w + 1) for _ in range(n + 1)]
#     for i in range(1,n+1):
#         for j in range(1,w+1):
#             KS[i][j]=max(KS[i-1][j-weights[i-1]]+values[i-1],KS[i-1][j]) if j>=weights[i-1] else KS[i-1][j]
#     return KS[n][w]
#     pass

File: Codes in C++/Dynamic Programming/0-1 Knapsack.cpp
int knapsack(vector<int> weight, vector<int> value, int n, int maxWeight) 
{
	vector<int> prev(maxWeight+1,0);
	for(int i=weight[0];i<=maxWeight;i++)prev[i]=value[0];
	for(int ind=1;ind<n;ind++)
	{
		for(int W=maxWeight;W>=0;W--)
		{
		int not_pick=0 + prev[W];
		int pick=INT_MIN;
		if(weight[ind]<=W)
			pick=value[ind]+prev[W-weight[ind]];
		prev[W]=max(pick,not_pick);
		}
	}
	
	return prev[maxWeight];
}


// int knapsack(vector<int> weight, vector<int> value, int n, int maxWeight) 
// {
// 	vector<int> prev(maxWeight+1,0);
// 	vector<int> cur(maxWeight+1,0);
// 	for(int i=weight[0];i<=maxWeight;i++)prev[i]=value[0];
// 	for(int ind=1;ind<n;ind++)
// 	{
// 		for(int W=0;W<=maxWeight;W++)
// 		{
// 		int not_pick=0 + prev[W];
// 		int pick=INT_MIN;
// 		if(weight[ind]<=W)
// 			pick=value[ind]+prev[W-weight[ind]];
// 		cur[W]=max(pick,not_pick);
// 		}
// 		prev=cur;
// 	}
	
// 	return prev[maxWeight];
// }



// int knapsack(vector<int> weight, vector<int> value, int n, int maxWeight) 
// {
// 	vector<vector<int>> dp(n,vector<int>(maxWeight+1,0));
// 	for(int i=weight[0];i<=maxWeight;i++)dp[0][i]=value[0];
// 	for(int ind=1;ind<n;ind++)
// 	{
// 		for(int W=0;W<=maxWeight;W++)
// 		{
// 		int not_pick=0 + dp[ind-1][W];
// 		int pick=INT_MIN;
// 		if(weight[ind]<=W)
// 			pick=value[ind]+dp[ind-1][W-weight[ind]];
// 		dp[ind][W]=max(pick,not_pick);
// 		}
// 	}
	
// 	return dp[n-1][maxWeight];
// }







// #include <bits/stdc++.h>
// int f(int ind,int maxWeight,vector<int> &weight,
//  vector<int> &value,vector<vector<int>> &dp)
// {
// 	if(ind==0)
// 	{
// 		if(weight[0]<=maxWeight)
// 			 return value[0];
// 		else 
// 			return 0;
// 	}

// 	if(dp[ind][maxWeight]!=-1) 
// 		return dp[ind][maxWeight];

// 	int not_pick=0 + f(ind-1,maxWeight,weight,value,dp);
// 	int pick=INT_MIN;
// 	if(weight[ind]<=maxWeight)
// 		pick=value[ind]+f(ind-1,maxWeight-weight[ind],weight,value,dp);
// 	dp[ind][maxWeight]=max(pick,not_pick);
// 	return dp[ind][maxWeight];
// }
// int knapsack(vector<int> weight, vector<int> value, int n, int maxWeight) 
// {
// 	vector<vector<int>> dp(n,vector<int>(maxWeight+1,-1));
	
// 	return f(n-1,maxWeight,weight,value,dp);
// }





// int maxProfit(vector<int> &values, vector<int> &weights, int n, int w)
// {
// 	int ks[n+1][w+1];
//     for (int i = 0; i < n + 1; i++)
// 		ks[i][0] = 0;
//  	for (int j = 0; j < w + 1; j++)
//         ks[0][j] = 0;

// 	for(int i=1;i<=n;i++)
// 	{
// 		for(int j=1;j<=w;j++)
// 		{
// 			int dont=ks[i-1][j];
// 			ks[i][j]=dont;
// 			if(j-weights[i-1]>=0)
// 			{
// 			int take=ks[i-1][j-weights[i-1]]+values[i-1];
// 			ks[i][j]=max(take,dont);

// 			}	 			
// 		}
// 	}
// 	return ks[n][w];
// }
